File: openrndr/src/main/kotlin/shape/FractalizedLine.kt
package shape

import org.openrndr.extra.noise.gaussian
import org.openrndr.extra.noise.random
import org.openrndr.math.Vector2
import org.openrndr.shape.Segment
import org.openrndr.shape.ShapeContour
import java.util.*
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/**
 * Based on the algorithm used here:
 * http://rectangleworld.com/blog/archives/462
 * and introduced here:
 * http://rectangleworld.com/blog/archives/413
 */

class FractalizedLine(var points: List<Vector2>, private val rng: Random = Random.Default) {
  constructor(other: FractalizedLine) : this(other.points, other.rng)

  val segments: List<Segment>
    get() {
      val segments = mutableListOf<Segment>()
      for (i in 0 until points.size - 1) {
        segments.add(Segment(points[i], points[i+ 1]))
      }
      // This is only needed if the path is closed, which this will never be.
      // Could be a possible enhancement in future: combine FractalizedPolygon and FractalizedLine with a `closed` boolean prop
      // segments.add(Segment(points.last(), points.first()))
      return segments
    }

  val shape: ShapeContour
    get() = ShapeContour(segments, true)

  /**
   * Recursively subdivide the points using perpendicular offset
   */
  fun perpendicularSubdivide(subdivisions: Int, offsetPct: Double = 0.50): FractalizedLine =
    subdivide(subdivisions, offsetPct, ::perpendicularOffset)

  /**
   * Recursively subdivide the points using gaussian offset
   * The default offsetPct here is intentionally lower than perpendicular because it gets _wild_ real quick
   */
  fun gaussianSubdivide(subdivisions: Int, offsetPct: Double = 0.35): FractalizedLine =
    subdivide(subdivisions, offsetPct, ::gaussianOffset)

  /**
   * Recursively subdivide the points
   */
  private fun subdivide(subdivisions: Int, offsetPct: Double = 0.5, offsetFn: (Vector2, Vector2, Double) -> Vector2): FractalizedLine {
    for (i in 0 until subdivisions) {
      val newPoints = mutableListOf<Vector2>()

      // Iterate through all points.
      // Skip the last index because we are accessing j+1 to get the "next" point anyway.
      // The last point will be added after the loop
      for (j in 0 until points.size - 1) {
        val current = points[j]
        val next = points[j + 1]
        val mid = offsetFn(current, next, offsetPct)
        newPoints.add(current)
        newPoints.add(mid)
      }
      newPoints.add(points.last())

      points = newPoints
    }
    return this
  }

  private fun perpendicularOffset(start: Vector2, end: Vector2, offsetPct: Double): Vector2 {
    val perpendicular = atan2(end.y - start.y, end.x - start.x) - (PI / 2.0)
    val maxDeviation = (start - end).length / 2.0 * offsetPct
    val mid = (start + end) / 2.0
    val offset = random(-maxDeviation, maxDeviation, rng)
    return mid + Vector2(cos(perpendicular) * offset, sin(perpendicular) * offset)
  }

  private fun gaussianOffset(start: Vector2, end: Vector2, offsetPct: Double): Vector2 =
    Vector2.gaussian(
      // midpoint
      mean = (start + end) / 2.0,
      deviation = Vector2((start - end).length / 2.0) * offsetPct,
      random = rng
    )

  fun clone(): FractalizedLine = FractalizedLine(this)
}

File: artsvg/lib/data-structures/fractalized-line.js
/**
 * Based on the algorithm used here:
 * http://rectangleworld.com/blog/archives/462
 * and introduced here:
 * http://rectangleworld.com/blog/archives/413
 *
 * Note: since the original algorithm has been removed:
 * It used a linked-list type structure which, yes, is more efficient, but
 * I chose to use a normal array because it is much, much easier to work with.
 * That said, I could potentially implement a custom iterator if I wanted to use a linked list in the future.
 */

import { LineSegment } from '../components/line-segment.js'
import { Path } from '../components/path.js'
import { random } from '../random.js'
import { Vector2 } from '../vector2.js'

/**
 * @class FractalizedLine
 * @classdesc Represents a fractalized line based on a given set of points.
 */
export class FractalizedLine {
  /**
   * @param {Vector2[]} points - The initial set of points.
   * @param {import("../random.js").Rng} [rng] - The random number generator. Defaults to the default Random instance.
   */
  constructor(points, rng = Math.random) {
    this.points = points
    this.rng = rng
  }

  /**
   * @readonly
   * @type {LineSegment[]} segments - The segments formed by connecting consecutive points.
   */
  get segments() {
    const segments = []
    for (let i = 0; i < this.points.length - 1; i++) {
      segments.push(new LineSegment(this.points[i], this.points[i + 1]))
    }
    return segments
  }

  /**
   * @param {boolean} closed
   * @returns {Path}
   */
  path(closed = true) {
    return Path.fromPoints(this.points, closed)
  }

  // TODO: openrndr/src/main/kotlin/shape/SmoothLine.kt
  // get smoothLine() {
  //   return new SmoothLine(this.points);
  // }

  /**
   * Recursively subdivide the points using perpendicular offset.
   * @param {number} subdivisions - The number of times to subdivide.
   * @param {number} [offsetPct=0.50] - The percentage of the offset.
   * @returns {FractalizedLine} - The updated FractalizedLine instance.
   */
  perpendicularSubdivide(subdivisions, offsetPct = 0.5) {
    return this.subdivide(
      subdivisions,
      offsetPct,
      this.perpendicularOffset.bind(this),
    )
  }

  // TODO: openrndr/src/main/kotlin/shape/FractalizedLine.kt
  // gaussianSubdivide(subdivisions, offsetPct = 0.35) {
  //   return this.subdivide(subdivisions, offsetPct, this.gaussianOffset);
  // }

  /**
   * Recursively subdivide the points.
   * @private
   * @param {number} subdivisions - The number of times to subdivide.
   * @param {number} [offsetPct=0.5] - The percentage of the offset.
   * @param {(v1: Vector2, v2: Vector2, n: number) => Vector2} [offsetFn=this.perpendicularOffset] - The offset function.
   * @returns {FractalizedLine} - The updated FractalizedLine instance.
   */
  subdivide(
    subdivisions,
    offsetPct = 0.5,
    offsetFn = this.perpendicularOffset.bind(this),
  ) {
    for (let i = 0; i < subdivisions; i++) {
      const newPoints = []

      for (let j = 0; j < this.points.length - 1; j++) {
        const current = this.points[j]
        const next = this.points[j + 1]
        const mid = offsetFn(current, next, offsetPct)
        newPoints.push(current, mid)
      }
      newPoints.push(this.points[this.points.length - 1])

      this.points = newPoints
    }
    return this
  }

  /**
   * Calculate the perpendicular offset between two points.
   * @private
   * @param {Vector2} start - The starting point.
   * @param {Vector2} end - The ending point.
   * @param {number} offsetPct - The percentage of the offset.
   * @returns {Vector2} - The calculated offset point.
   */
  perpendicularOffset(start, end, offsetPct) {
    const perpendicular =
      Math.atan2(end.y - start.y, end.x - start.x) - Math.PI / 2.0
    const maxDeviation = (start.subtract(end).length() / 2.0) * offsetPct
    const mid = start.add(end).div(2)
    const offset = random(-maxDeviation, maxDeviation, this.rng)
    return new Vector2(
      mid.x + Math.cos(perpendicular) * offset,
      mid.y + Math.sin(perpendicular) * offset,
    )
  }
}
